Refactor ShutdownResult type and construction

wvanlint opened the pull request:
As a follow-up to https://github.com/lightningdevkit/rust-lightning/pull/2486, this PR:
- Refactors the `ShutdownResult` type and moves construction to `Channel` methods.
- Cleans up typos and unused variables/imports.
- Refactors `check_closed_event` to be able to share it for batch funding tests. The current implementation is not usable since there might be multiple `ClosureReason`s. Let me know if this is an overgeneralization though.

wvanlint commented on lightning/src/ln/channel.rs:
```diff
@@ -5557,7 +5573,7 @@ impl<SP: Deref> Channel<SP> where
 	pub fn get_shutdown(&mut self, signer_provider: &SP, their_features: &InitFeatures,
 		target_feerate_sats_per_kw: Option<u32>, override_shutdown_script: Option<ShutdownScript>)
-	-> Result<(msgs::Shutdown, Option<ChannelMonitorUpdate>, Vec<(HTLCSource, PaymentHash)>), APIError>
+	-> Result<(msgs::Shutdown, Option<ChannelMonitorUpdate>, Vec<(HTLCSource, PaymentHash)>, Option<ShutdownResult>), APIError>
```
I wonder if these `Option<ChannelMonitorUpdate>, Vec<(HTLCSource, PaymentHash)>` can be collapsed into `Option<ShutdownResult>` although the use cases are slightly different.

wvanlint replied:
Right, that was my main hesitation. It seems possible to branch on the `update_id` of the `ChannelMonitorUpdate` being equal to `CLOSED_CHANNEL_UPDATE_ID`, but might not be something we want to unify.

wvanlint commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6972,8 +6966,7 @@ where
 		let mut has_update = false;
-		let mut shutdown_result = None;
-		let mut unbroadcasted_batch_funding_txid = None;
+		let mut shutdown_results = Vec::new();
```
This is a correction I believe @TheBlueMatt @wpaulino.

wvanlint commented:
> Any reason this is marked as draft?

Took another look, ready for review now. :)

benthecarman commented on lightning/src/ln/channel.rs:
```diff
@@ -4207,18 +4209,18 @@ impl<SP: Deref> Channel<SP> where
 	pub fn maybe_propose_closing_signed<F: Deref, L: Deref>(
 		&mut self, fee_estimator: &LowerBoundedFeeEstimator<F>, logger: &L)
-		-> Result<(Option<msgs::ClosingSigned>, Option<Transaction>), ChannelError>
+		-> Result<(Option<msgs::ClosingSigned>, Option<Transaction>, Option<ShutdownResult>), ChannelError>
```
This was added to the return but it looks like it's always none

vladimirfomene commented on lightning/src/ln/channel.rs:
```diff
@@ -5626,11 +5642,19 @@ impl<SP: Deref> Channel<SP> where
+		let shutdown_result;
+		let unbroadcasted_batch_funding_txid = self.context.unbroadcasted_batch_funding_txid();
 		self.context.target_closing_feerate_sats_per_kw = target_feerate_sats_per_kw;
 		if self.context.channel_state & !STATE_FLAGS < ChannelState::FundingSent as u32 {
```
Instead of defining `shutdown_result` on L5645 won't it be better to have this if/else expression return a `ShutdownResult` and then capture it in the `shutdown_result` variable.

wvanlint replied:
Done, capturing `shutdown_result` from if-else expressions in a couple places now.

vladimirfomene reviewed (commented):
Just a couple of comments here and there

wvanlint commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1467,27 +1467,66 @@ macro_rules! check_closed_broadcast {
+	pub reason: Option<ClosureReason>,
+}
+
+impl Default for ExpectedCloseEvent {
```
You're right, derived instead. :)

wvanlint commented:
> Can you write more full commit messages that describe what we're doing (not just "refactor", but "refactor to do X") and why we're doing it?
I've rewritten the commit subjects and added explanations to the bodies, let me know if anything can be cleaned up there though!

wvanlint commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1467,27 +1467,55 @@ macro_rules! check_closed_broadcast {
+				channel_capacity_sats,
+				..
+			} if (
+				expected_event.channel_id.map(|expected| *channel_id == expected).unwrap_or(true) &&
```
The `.map` function is called on the expected event field, so the `unwrap_or(true)` call is meant to match if the caller left expectations unspecified. Does that sound correct? A `ChannelClosed` event will always have a `channel_id`.

wvanlint commented on lightning/src/ln/channel.rs:
```diff
@@ -542,18 +542,16 @@ pub(super) struct ReestablishResponses {
-	Option<Txid>
-);
+/// The result of a shutdown that should be handled.
+pub(crate) struct ShutdownResult {
```
Done.

wvanlint commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2574,10 +2574,10 @@ where
+						let (shutdown_msg, mut monitor_update_opt, htlcs, local_shutdown_result) =
 							chan.get_shutdown(&self.signer_provider, their_features, target_feerate_sats_per_1000_weight, override_shutdown_script)?;
 						failed_htlcs = htlcs;
+						shutdown_result = local_shutdown_result;
```
Added this assertion to all the cooperative closing paths.

wvanlint commented on lightning/src/ln/functional_test_utils.rs:
```diff
@@ -1467,27 +1467,55 @@ macro_rules! check_closed_broadcast {
 /// Check that a channel's closing channel events has been issued
-pub fn check_closed_event(node: &Node, events_count: usize, expected_reason: ClosureReason, is_check_discard_funding: bool,
+pub fn check_closed_event(node: &Node, _events_count: usize, expected_reason: ClosureReason, is_check_discard_funding: bool,
 	expected_counterparty_node_ids: &[PublicKey], expected_channel_capacity: u64) {
```
Needed to account for the `DiscardFunding` events too but verified the `events_count` here.

wvanlint commented:
> Ugh, sorry, this needs rebase due to a silent conflict with #2658.

No problem! Rebased. :)
